x-pack/plugins/event_log/server/event_log_client.ts

```diff
@@ -138,6 +138,26 @@ export class EventLogClient implements IEventLogClient {
     });
   }
 
+  public async aggregateEventsBySavedObjectType(type: string, options?: AggregateOptionsType) {
```

Comment:
I tried this approach initially and I wasn't able to figure out where to place this filter in the query. Most obvious place was [here](https://github.com/elastic/kibana/blob/main/x-pack/plugins/event_log/server/es/cluster_client_adapter.ts#L449) in `cluster_client_adapter.ts`:

```js
let dslFilterQuery: estypes.QueryDslBoolQuery['filter'];
  try {
    dslFilterQuery = filter ? toElasticsearchQuery(fromKueryExpression(filter)) : undefined;
  }
```

but merging the `authFilter` into the `filter` seems to just filter out everything.

Could use some advice here on how to proceed.

Reply:
@Zacqary This seemed to work for me:

In `x-pack/plugins/event_log/server/es/cluster_client_adapter.ts` leave the existing functions untouched and add a new function:

```
export interface AggregateEventsWithAuthFilter {
  index: string;
  namespace: string | undefined;
  type: string;
  authFilter: KueryNode;
  legacyIds?: string[];
  aggregateOptions: AggregateOptionsType;
}

public async aggregateEventsWithAuthFilter(
    queryOptions: AggregateEventsWithAuthFilter
  ): Promise<AggregateEventsBySavedObjectResult> {
    const { index, type, aggregateOptions } = queryOptions;
    const { aggs } = aggregateOptions;

    const esClient = await this.elasticsearchClientPromise;

    const query = getQueryBodyWithAuthFilter(
      this.logger,
      queryOptions,
      pick(queryOptions.aggregateOptions, ['start', 'end', 'filter'])
    );

    const body: estypes.SearchRequest['body'] = {
      size: 0,
      query,
      aggs,
    };

    try {
      const { aggregations } = await esClient.search<IValidatedEvent>({
        index,
        body,
      });
      return {
        aggregations,
      };
    } catch (err) {
      throw new Error(`querying for Event Log by for type "${type}" and auth filter failed with: ${err.message}`);
    }
  }
}

export function getQueryBodyWithAuthFilter(
  logger: Logger,
  opts: AggregateEventsWithAuthFilter,
  queryOptions: QueryOptionsType
) {
  const { namespace, type, authFilter } = opts;
  const { start, end, filter } = queryOptions ?? {};

  const namespaceQuery = getNamespaceQuery(namespace);
  let dslFilterQuery: estypes.QueryDslBoolQuery['filter'];
  try {
    const filterKueryNode = filter ? fromKueryExpression(filter) : null;
    const queryFilter = filterKueryNode
      ? nodeBuilder.and([filterKueryNode, authFilter as KueryNode])
      : authFilter;
    dslFilterQuery = queryFilter ? toElasticsearchQuery(queryFilter) : undefined;
  } catch (err) {
    logger.debug(
      `esContext: Invalid kuery syntax for the filter (${filter}) error: ${JSON.stringify({
        message: err.message,
        statusCode: err.statusCode,
      })}`
    );
    throw err;
  }

  const savedObjectsQueryMust: estypes.QueryDslQueryContainer[] = [
    {
      term: {
        'kibana.saved_objects.rel': {
          value: SAVED_OBJECT_REL_PRIMARY,
        },
      },
    },
    {
      term: {
        'kibana.saved_objects.type': {
          value: type,
        },
      },
    },
    // @ts-expect-error undefined is not assignable as QueryDslTermQuery value
    namespaceQuery,
  ];

  const musts: estypes.QueryDslQueryContainer[] = [
    {
      nested: {
        path: 'kibana.saved_objects',
        query: {
          bool: {
            must: reject(savedObjectsQueryMust, isUndefined),
          },
        },
      },
    },
  ];

  if (start) {
    musts.push({
      range: {
        '@timestamp': {
          gte: start,
        },
      },
    });
  }
  if (end) {
    musts.push({
      range: {
        '@timestamp': {
          lte: end,
        },
      },
    });
  }

  return {
    bool: {
      ...(dslFilterQuery ? { filter: dslFilterQuery } : {}),
      must: reject(musts, isUndefined),
    },
  };
}
```

In `x-pack/plugins/event_log/server/event_log_client.ts`, add this function:

```
public async aggregateEventsWithAuthFilter(
    type: string,
    authFilter: KueryNode,
    options?: AggregateOptionsType
  ) {
    if (!authFilter) {
      throw new Error('No authorization filter defined!');
    }

    const aggs = options?.aggs;
    if (!aggs) {
      throw new Error('No aggregation defined!');
    }

    // validate other query options separately from
    const aggregateOptions = queryOptionsSchema.validate(omit(options, 'aggs') ?? {});

    return await this.esContext.esAdapter.aggregateEventsWithAuthFilter({
      index: this.esContext.esNames.indexPattern,
      namespace: await this.getNamespace(),
      type,
      authFilter,
      aggregateOptions: { ...aggregateOptions, aggs } as AggregateOptionsType,
    });
  }
```

and use inside `rulesClient.getGlobalExecutionLogWithAuth`

```
const aggResult = await eventLogClient.aggregateEventsWithAuthFilter(
        'alert',
        authorizationTuple.filter as KueryNode,
        {
          start: parsedDateStart.toISOString(),
          end: parsedDateEnd.toISOString(),
          aggs: getExecutionLogAggregation({
            filter,
            page,
            perPage,
            sort,
          }),
      });
```

Reply:
I think it's important to separate out the functions inside the cluster client adapter since one takes an array of IDs as its "auth" and the other should take an explicit "authFilter". I don't think we should be passing in a `getAllIds` flag to override passing in an empty list of IDs.

There are probably ways of combining `getQueryBody` and `getQueryBodyWithAuthFilter` but I didn't look too hard at doing that.

Reply:
LMK if that works for you!

Reply:
@ymao1 I've copied all this and separated out all the functions but it's still just filtering everything out and not matching anything at all. This authorization filter only seems to work when applied to the aggregation.

Might need to pair with you on this to figure out what's going on.